lib/ansible/modules/cloud/amazon/s3.py

```diff
@@ -272,131 +240,150 @@
-                exception=traceback.format_exc())
+            s3.put_bucket_acl(ACL=acl, Bucket=bucket)
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Failed while creating bucket or setting acl (check that you have CreateBucket and PutBucketAcl permission).",
+                         exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+
     if bucket:
         return True
 
-def get_bucket(module, s3, bucket):
-    try:
-        return s3.lookup(bucket)
-    except s3.provider.storage_response_error as e:
-        module.fail_json(msg="Failed while getting bucket %s: %s" % (bucket, e),
-                exception=traceback.format_exc())
 
-def list_keys(module, bucket_object, prefix, marker, max_keys):
-    all_keys = bucket_object.get_all_keys(prefix=prefix, marker=marker, max_keys=max_keys)
+def paginated_list(s3, **pagination_params):
+    pg = s3.get_paginator('list_objects_v2')
+    for page in pg.paginate(**pagination_params):
+        for data in page.get('Contents', {}):
+            yield data['Key']
 
-    keys = [x.key for x in all_keys]
 
-    module.exit_json(msg="LIST operation complete", s3_keys=keys)
+def list_keys(module, s3, bucket, prefix, marker, max_keys):
+    pagination_params = {'Bucket': bucket}
+    for param_name, param_value in (('Prefix', prefix), ('StartAfter', marker), ('MaxKeys', max_keys)):
+        pagination_params[param_name] = param_value
+    try:
+        keys = [key for key in paginated_list(s3, **pagination_params)]
+        module.exit_json(msg="LIST operation complete", s3_keys=keys)
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Failed while listing the keys in the bucket {0}".format(bucket),
+                         exception=traceback.format_exc(),
+                         **camel_dict_to_snake_dict(e.response))
+
 
 def delete_bucket(module, s3, bucket):
     if module.check_mode:
         module.exit_json(msg="DELETE operation skipped - running in check mode", changed=True)
     try:
-        bucket = s3.lookup(bucket, validate=False)
-        bucket_contents = bucket.list()
-        bucket.delete_keys([key.name for key in bucket_contents])
-    except s3.provider.storage_response_error as e:
-        if e.status == 404:
-            # bucket doesn't appear to exist
+        exists = bucket_check(module, s3, bucket)
+        if exists is False:
             return False
-        elif e.status == 403:
-            # bucket appears to exist but user doesn't have list bucket permission; may still be able to delete bucket
-            pass
-        else:
-            module.fail_json(msg=str(e), exception=traceback.format_exc())
-    try:
-        bucket.delete()
+        # if there are contents then we need to delete them before we can delete the bucket
+        keys = [{'Key': key} for key in paginated_list(s3, **{'Bucket': bucket})]
```

Comment:
`**{'Bucket': bucket}` is equivalent to `Bucket=bucket`. Please use the latter :)